### ISSUE TITLE
unpickleable exceptions cause queue to crash

### ISSUE THREAD

**Issue by caleb15** [celery/celery#6185]
# Checklist

* [x] I have verified that the issue exists against the `master` branch of Celery. (not against latest commit specifically but against latest version 4.4.5)
* [x] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
* [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
* [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
* [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
* [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
* [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information
* [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected). 
* [x] I have verified that the issue exists against the `master` branch of Celery.
* [x] I have included the contents of ``pip freeze`` in the issue.
* [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

* [x] I have tried reproducing the issue on more than one Python version
  and/or implementation.
* [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
* [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
* [ ] I have tried reproducing the issue on more than one operating system.
* [ ] I have tried reproducing the issue on more than one workers pool.
* [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
* [x] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues
* None

#### Possible Duplicates
* None

## Environment & Settings

**Celery version**: 4.4.5 (cliffs)

Celery Report:
https://pastebin.com/0x13Xjiq

# Steps to Reproduce
## Required Dependencies

* **Minimal Python Version**: 3.6.8
* **Minimal Celery Version**: 4.4.1
* **Minimal Kombu Version**: 4.6.8
* **Minimal Broker Version**: Unknown
* **Minimal Result Backend Version**: Unknown
* **Minimal OS and/or Kernel Version**: Ubuntu 18
* **Minimal Broker Client Version**: Unknown
* **Minimal Result Backend Client Version**: Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
Only relevant packages shown below as 15five is not open-source:
amqp==2.6.0
billiard==3.6.3.0
celery==4.4.5
Django==2.2.13
django-celery-results==1.2.1
future==0.18.2
importlib-metadata==1.6.1
kombu==4.6.10
vine==1.3.0
pytz==2020.1
zipp==3.1.0
```

</p>
</details>

### Other Dependencies

<details>
<p>
broker: AWS elasticache redis
</p>
</details>

## Minimally Reproducible Test Case

https://github.com/caleb15/celeryPickleBugReproduction

# Expected Behavior
Error to not happen. If error is a result of our faulty user code, then I would expect the celery queue to log a error and continue operating.

# Actual Behavior

We get a AttributeError and the queue dies. Upon celery restart the queue continues to process events, but eventually hit the same error and dies again.

I can send a link to the sentry error event directly to a celery engineer if one is interested. Otherwise the relevant info is below:

```
AttributeError billiard.reduction in loadbuf
Error on stopping Pool: "AttributeError(\"can't set attribute\")"
billiard/reduction.py in loadbuf at line 61
billiard/connection.py in recv at line 279
celery/concurrency/asynpool.py in _flush_outqueue at line 412
celery/concurrency/asynpool.py in iterate_file_descriptors_safely at line 240
celery/concurrency/asynpool.py in on_stop_not_started at line 387
billiard/pool.py in stop at line 526
billiard/pool.py in stop_if_not_current at line 145
billiard/pool.py in join at line 1622
celery/concurrency/prefork.py in on_stop at line 145
celery/concurrency/base.py in stop at line 123
celery/bootsteps.py in stop at line 373
celery/bootsteps.py in send_all at line 151
```

Another stacktrace:

```
celery/concurrency/asynpool.py in _recv_message at line 332
celery/concurrency/asynpool.py in on_result_readable at line 351
kombu/asynchronous/hub.py in create_loop at line 364
celery/worker/loops.py in asynloop at line 83
celery/worker/consumer/consumer.py in start at line 599
celery/bootsteps.py in start at line 119
celery/worker/consumer/consumer.py in start at line 318
celery/bootsteps.py in start at line 369
celery/bootsteps.py in start at line 119
celery/worker/worker.py in start at line 208
```

sys.argv: 

```
[/srv/venvs/py3/lib/python3.7/site-packages/celery/__main__.py, worker, -A, ff, --loglevel=INFO, -Q, celery, --logfile=/var/log/celery/default%I.log, --pidfile=/var/run/celery/default.pid, --hostname=default@celery0]
```

## What we have tried so far:
We have tried the following without success:

* Updating python version from 3.6.8 to 3.7.7
* Upgrading celery
* Moving most tasks out of the default queue into other queues
* pinpointing which task caused the error by looking through logs (we have been unable to pinpoint any specific task, the last task processed before error varies)

## [update] identification & cause of error:
Our staff engineer noticed the task responsible in the sentry breadcrumbs and I figured out that the task was raising a ValidationError that I suspected was not pickleable. We ignored the error and so far the queue hasn't died on us. My suspicion was confirmed when a CI failure revealed that the error was unpickleable:

```
test_update_billing_add_ons_ignores_simultaneous_request_error (ff.mods.billing.tests.test_subscription_logic.UpdateBillingAddOnsTestCase) failed:

    ValidationError(b'<?xml version="1.0" encoding="UTF-8"?>\n<error>\n
    <symbol>simultaneous_request</symbol>\n  <description>A change for
    subscription 456ee8a3881160536ad68346fe91eede is already in
    progress.</description>\n</error>')

Unfortunately, the exception it raised cannot be pickled, making it impossible
for the parallel test runner to handle it cleanly.

Here's the error encountered while trying to pickle the exception:

    AttributeError("can't set attribute")

You should re-run this test with the --parallel=1 option to reproduce the
failure and get a correct traceback.

Destroying test database for alias 'default'...
TOTAL RUNTIME: 67.09s
TOTAL RUNTIME: 67.09s
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/runner.py", line 309, in _run_subsuite
    result = runner.run(subsuite)
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/runner.py", line 256, in run
    test(result)
  File "/usr/local/lib/python3.7/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/lib/python3.7/unittest/suite.py", line 122, in run
    test(result)
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/testcases.py", line 271, in __call__
    super().__call__(result)
  File "/usr/local/lib/python3.7/unittest/case.py", line 676, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/lib/python3.7/unittest/case.py", line 636, in run
    self._feedErrorsToResult(result, outcome.errors)
  File "/usr/local/lib/python3.7/unittest/case.py", line 566, in _feedErrorsToResult
    result.addError(test, exc_info)
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/runner.py", line 197, in addError
    self.check_picklable(test, err)
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/runner.py", line 130, in check_picklable
    self._confirm_picklable(err)
  File "/home/circleci/project/venv/lib/python3.7/site-packages/django/test/runner.py", line 104, in _confirm_picklable
    pickle.loads(pickle.dumps(obj))
AttributeError: can't set attribute
```

**Comment by caleb15** [comment]
**TL;DR:**

Unpickleable errors can cause the celery queue to crash. The queue should log a warning and continue operating.

**Comment by auvipy** [comment]
are you up dor a fix?

**Comment by caleb15** [comment]
Probably not. I'll create a ticket for fixing it in our jira instance but I'm not sure when I'll have time to get around to it. I would need to figure out how to easily reproduce it first.

Can we split the work? I give you code to reproduce the error, you do the fix? :pray:

**Comment by auvipy** [comment]
please provide the test

**Comment by georgepsarakis** [comment]
Hello @caleb15 . Normally unpickleable exceptions should not be an issue, since there is specific handling for them. For details see [here](https://github.com/celery/celery/blob/f653dce4557c06c8d6f933f7c43486897747abff/celery/utils/serialization.py#L169). If you could provide a test case to reproduce the problem, that would be great indeed.

Just noting, that if you want to handle the exception in your task and reject the message, this is also possible by raising [celery.exceptions.Reject](https://github.com/celery/celery/blob/bf6139bf651b20bc04b895a5f6eb8d50320bc252/celery/app/task.py#L723).

**Comment by caleb15** [comment]
I've been looking into this a bit more but haven't been able to replicate just yet.

Quick question: why is celery using pickle? I looked in the docs and found out that it switched to using json as the serializer since v4. I looked at my configuration and confirmed that we don't configure [task_serializer](https://docs.celeryproject.org/en/stable/userguide/configuration.html?highlight=pickle#task-serializer) so it must be using the default json serializer.

**Comment by thedrow** [comment]
Because that's the only way to properly represent the exception & the traceback in serialized form, I think.

**Comment by caleb15** [comment]
Here is a simple repo I made for reproducing the issue:

https://github.com/caleb15/celeryPickleBugReproduction

I included a readme with quick replication steps and the error stacktrace :)

**Comment by thedrow** [comment]
Thank you.
I wonder if it will crash in our integration suite either.
If you don't mind to find out, please try.

If not, I'll get to it.

**Comment by auvipy** [comment]
the code I saw use XML

**Comment by thedrow** [comment]
I can't reproduce this with 3.8.5 and 3.9.0.

**Comment by caleb15** [comment]
versions 3.8.5 and 3.9.0 of what?

**Comment by thedrow** [comment]
Yes.

**Comment by caleb15** [comment]
Are you able to reproduce it with python 3.7.7?

**Comment by thedrow** [comment]
I tried pickling and unpickling the same exception with 3.7.10. It works.

**Comment by thedrow** [comment]
```
sudo docker run -it python:3.7.7 /bin/bash
root@697175da4188:/# pip install recurly==2.9.16
Collecting recurly==2.9.16
  Downloading recurly-2.9.16-py3-none-any.whl (26 kB)
Collecting defusedxml
  Downloading defusedxml-0.7.1-py2.py3-none-any.whl (25 kB)
Collecting six>=1.4.0
  Downloading six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting iso8601
  Downloading iso8601-0.1.14-py2.py3-none-any.whl (9.5 kB)
Collecting backports.ssl-match-hostname
  Downloading backports.ssl_match_hostname-3.7.0.1.tar.gz (5.7 kB)
Building wheels for collected packages: backports.ssl-match-hostname
  Building wheel for backports.ssl-match-hostname (setup.py) ... done
  Created wheel for backports.ssl-match-hostname: filename=backports.ssl_match_hostname-3.7.0.1-py2.py3-none-any.whl size=6741 sha256=f9e888c1f3d0f6244ecbb053d8a7ac406601a717784aa2a665319a96ae2fec8e
  Stored in directory: /root/.cache/pip/wheels/47/70/47/fa57137a308c61c2891887ba874eec1e0252a857448124896c
Successfully built backports.ssl-match-hostname
Installing collected packages: defusedxml, six, iso8601, backports.ssl-match-hostname, recurly
Successfully installed backports.ssl-match-hostname-3.7.0.1 defusedxml-0.7.1 iso8601-0.1.14 recurly-2.9.16 six-1.15.0
WARNING: You are using pip version 20.1.1; however, version 21.0.1 is available.
You should consider upgrading via the '/usr/local/bin/python -m pip install --upgrade pip' command.
root@697175da4188:/# python
Python 3.7.7 (default, Jun  9 2020, 17:58:51)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pickle import *
>>> from recurly.errors import ValidationError
>>> some_xml = b"""<?xml version="1.0" encoding="UTF-8"?>
... <error>
...   <symbol>simultaneous_request</symbol>
...   <description>A change for subscription 456ee8a3881160536ad68346fe91eede is already in progress.</description>
... </error>"""
>>> loads(dumps(ValidationError(some_xml))
... )
ValidationError(b'<?xml version="1.0" encoding="UTF-8"?>\n<error>\n  <symbol>simultaneous_request</symbol>\n  <description>A change for subscription 456ee8a3881160536ad68346fe91eede is already in progress.</description>\n</error>')
```

🤷

**Comment by thedrow** [comment]
@auvipy Even if the exception can be pickled, this might still be reproducible if we raise an AttributeError manually at the appropriate place.

**Comment by caleb15** [comment]
Yep, for reproducing please use https://github.com/caleb15/celeryPickleBugReproduction

**Comment by thedrow** [comment]
But I can pickle and unpickle the exception.
How is this any different in Celery?

**Comment by caleb15** [comment]
> How is this any different in Celery?

Not sure, but celery fails so something must be going on :man_shrugging:

**Comment by auvipy** [comment]
I think you should try XML serializer of django. I am not sure xml is rightly encoded as you are trying in your code.

**Comment by caleb15** [comment]
Thanks for the tip, but we already avoided the issue by logging the exception instead of raising it. Even if Celery finds invalid XML in the task exception, it should log an error instead of crashing the queue.

**Comment by auvipy** [comment]
> site-packages/django/test/runner.py", line 104, in _confirm_picklable
>     pickle.loads(pickle.dumps(obj))
> AttributeError: can't set attribute

yes, for that reason, we still keep the issue open, but you were facing trouble for not having right coding approach.

**Comment by caleb15** [comment]
Fair enough. I'm not the one who coded that so I don't take any offense :D

**Comment by auvipy** [comment]
I am not offending any one, just trying to help to figure out the approach.

**Comment by taleinat** [comment]
> Because that's the only way to properly represent the exception & the traceback in serialized form, I think.

Might want to take a look at [tblib](https://pypi.org/project/tblib/), which can (de)serialize tracebacks using pickle and at least two other formats.

**Comment by youtux** [comment]
Hi, what's the status of this? I still see the issue with python 3.10 and celery 5.2.7 (and 5.3.0b2). Do you need an MCVE, or is it clear how to reproduce the bug?

**Comment by youtux** [comment]
To me, it seems that it only happens when using retry logic with an unpicklable exception (e.g. doing a `self.retry(MyUnpicklableException())`).
If I just `raise MyUnpicklableException()`, then it is wrapped correctly and the worker is able to recover just fine.

**Comment by youtux** [comment]
If it can be of help, I fixed it in one of the codebases I manage by adding this code to my `celery_app.py` definition:

```python
# celery_app.py

app = Celery(...)

@task_retry.connect
def on_task_retry(*args, einfo, **kwargs):
    """Make sure that exceptions passed to the `retry` method are wrapped in a
    UnpickleableExceptionWrapper if needed.

    This is a part of the fix for the celery bug that causes the worker to
    crashloop when a task raises a `Retry` with a non "pickle-unpicklable" exception.

    Celery (up to version 5.3.0b2) already does this for exceptions raised by a task
    but that are not connected to a retry of the task, but for some reason they didn't
    apply this logic to the handling of exceptions passed to the `retry` method.

    Bug upstream: https://github.com/celery/celery/issues/6185
    """
    einfo.exception = get_pickleable_exception(einfo.exception)
    einfo.type = get_pickleable_etype(einfo.type)


def patch_celery_exception_wrapper():
    """Monkey patch celery's UnpickleableExceptionWrapper to make sure that it handles
    arguments that can't surivive a pickle roundtrip.

    This is a part of the fix for the celery bug that causes the worker to
    crashloop when a task raises a `Retry` with a non "pickle-unpicklable" exception.

    Celery versions at least up to 5.3.0b2 suffer from a bug in the way it wraps
    exceptions whoese arguments that are not "pickle-unpickleable" (i.e. items that
    can't survive a pickle roundtrip). With such arguments,
    `UnpickleableExceptionWrapper.__init__` will not correclty escape them, since it
    uses only `pickle.dumps(x)` to check if the exception is pickleable.
    The correct way is to check the roundtrip with `pickle.loads(pickle.dumps(x))`.

    The side effect of this bug is that the celery worker will go in a crashloop
    where it tries to `pickle.loads` the exception.

    This problem usually happens when a task rises a `Retry` with a non
    "pickle-unpicklable" exception as argument.
    For example, the following task will cause the worker to crashloop (`SlackApiError`
    from slackclient==2.9.4 is non-unpicklable):
    >>> @app.task(bind=True)
    ... def my_task(self):
    ...     self.retry(exc=slack.errors.SlackApiError("foo", "bar"), countdown=1)
    """
    orig_init = UnpickleableExceptionWrapper.__init__

    def new_init(self, exc_module, exc_cls_name, exc_args, text=None):
        safe_exc_args = ensure_serializable(
            exc_args, lambda v: pickle.loads(pickle.dumps(v))
        )
        orig_init(self, exc_module, exc_cls_name, safe_exc_args, text)

    new_init._original_init = orig_init

    UnpickleableExceptionWrapper.__init__ = new_init
    return orig_init


patch_celery_exception_wrapper()
```

**Comment by youtux** [comment]
Is this issue a duplicate of https://github.com/celery/celery/issues/6990? If so, shall we close it?

**Comment by auvipy** [comment]
> Is this issue a duplicate of #6990? If so, shall we close it?

did you check this comment https://github.com/celery/celery/issues/6990#issuecomment-1465200272 ? is the same issue causing the error?

**Comment by youtux** [comment]
I don't think so, as the issue for me started with the celery 5.1.0 release (5.0.5 is the latest release that is not affected).
This is the MVCE I'm using to check the issue:

```
# tasks.py

from celery import Celery

app = Celery("tasks", broker="redis://localhost:6379/0", backend="redis://localhost:6379/0")

class BadException(Exception):
    def __init__(self, message, code):
        super().__init__(message)
        self.code = code


@app.task(bind=True)
def add(self):
    raise self.retry(exc=BadException("foo", "bar"), countdown=1, max_retries=1)
```

```
# run celery
celery --app tasks worker --loglevel DEBUG
```

```
# run the task, it will never get a result, nor an exception, as the worker gets stuck
python -c "from tasks import add; add.delay().get()"
```
